internal/service/alarm_subscription_handler.go

```diff
@@ -358,3 +402,206 @@ func (h *alarmSubscriptionHandler) mapItem(ctx context.Context,
+
+	//apply new data
+
+	//u := &unstructured.Unstructured{}
+	//u.Object, _ = runtime.DefaultUnstructuredConverter.ToUnstructured(configmap)
+	err = h.hubClient.Patch(ctx, configmap, clnt.Apply, clnt.ForceOwnership, clnt.FieldOwner("field-owner"))
+
+	return err
+}
+
+func (h *alarmSubscriptionHandler) ConfigMapReadEntry(
+	ctx context.Context, entry_key string) (value string, err error) {
+
+	//test to read the configmap
+	configmap := &corev1.ConfigMap{}
+
+	key := clnt.ObjectKey{
+		Namespace: TESTNAMESPACE,
+		Name:      TESTCONFIGMAPNAME,
+	}
+	err = h.hubClient.Get(ctx, key, configmap)
+
+	if err != nil && !apierrors.IsNotFound(err) {
+		//there is error and err is not notfound error
+		//panic("unexpected error")
+		return
+	}
+
+	if configmap.Data == nil {
+		return
+	}
+
+	value, ok := configmap.Data[entry_key]
+	if !ok {
+		err = ErrNotFound
+	}
+
+	return
+}
+
+func (h *alarmSubscriptionHandler) ConfigMapDeleteEntry(
+	ctx context.Context, entry_key string) (err error) {
+
+	//test to read the configmap
+	configmap := &corev1.ConfigMap{}
+
+	key := clnt.ObjectKey{
+		Namespace: TESTNAMESPACE,
+		Name:      TESTCONFIGMAPNAME,
+	}
+	err = h.hubClient.Get(ctx, key, configmap)
+
+	if err != nil && !apierrors.IsNotFound(err) {
+		//there is error and err is not notfound error
+		//panic("unexpected error")
+		return
+	}
+
+	if configmap.Data == nil {
+		return
+	}
+
+	_, ok := configmap.Data[entry_key]
+
+	//entry not found
+	if !ok {
+		return
+	}
+	delete(configmap.Data, entry_key)
+
+	// Note: there is a case configmap is empty shall we remove the configmap
+	err = h.hubClient.Update(ctx, configmap, &clnt.UpdateOptions{})
+
+	return
+}
+
+func (h *alarmSubscriptionHandler) RecoveryFromConfigMap(ctx context.Context) (err error) {
+
+	//test to read the configmap
+	configmap := &corev1.ConfigMap{}
+
+	key := clnt.ObjectKey{
+		Namespace: TESTNAMESPACE,
+		Name:      TESTCONFIGMAPNAME,
+	}
+	err = h.hubClient.Get(ctx, key, configmap)
+
+	if err != nil && !apierrors.IsNotFound(err) {
+		//there is error and err is not notfound error
+		//panic("unexpected error")
+		return
+	}
+
+	if configmap.Data == nil {
+		return
+	}
+
+	for mapkey, value := range configmap.Data {
+		var object data.Object
+		err = jsoniter.Unmarshal([]byte(value), &object)
+		if err != nil {
+			continue
+		}
+
+		h.subscriptionMap[mapkey] = object
+
+	}
+
+	return
+}
+
+func (h *alarmSubscriptionHandler) WatchConfigMap(ctx context.Context) (err error) {
+
+	raw_opt := metav1.SingleObject(metav1.ObjectMeta{Name: TESTCONFIGMAPNAME, Namespace: TESTNAMESPACE})
+	opt := clnt.ListOptions{}
+	opt.Raw = &raw_opt
+
+	watcher, err := h.hubClient.Watch(ctx, &corev1.ConfigMapList{}, &opt)
+
+	if err != nil {
+		panic("failed to launch watcher")
+	}
+
+	go func() {
+		for {
+			event, open := <-watcher.ResultChan()
+			if open {
+				switch event.Type {
+				case watch.Added:
+					fallthrough
+				case watch.Modified:
+					h.subscritionMapMemoryLock.Lock()
```

Comment:
When you acquire a lock it is good practice to ensure that it will be released even if there are errors. Here, for example, you have a `continue` statement that will end the iteration without releasing the lock. One possible solution for that is to create an anonymous function and use the `defer` mechanism:

```go
func () {
    h.subscriptionMapMemoryLock.Lock()
    defer h.subscriptionMemoryLock.Unlock()
    ...
}()
```

You can also use a method instead of an anonymous function, if there are not many local variables to pass.

Reply:
Changed.